### PR TITLE
feat(gemini): autogenerated update

### PR DESCRIPTION
**Pull request by polsala**
## Rationale
- Automated proposal from the Gemini generator delivering a fresh community utility.

## Why safe to merge
- Utility is isolated to `utils/<util_name>`.
- README + tests ship together (see folder contents).
- No secrets or credentials touched.

## Test Plan
- Follow the instructions in the generated README

## Links
- Generated docs and examples committed alongside this change.

## Mock Justification
- Not applicable; generator did not introduce new mocks.

### CODE REVIEW

**Comment by polsala** [comment]
✅ What’s solid
*   **Modular Design:** The separation of concerns into `Recipe`, `SurvivalCookbook`, and `recommend_recipes` classes/functions promotes readability, maintainability, and testability.
*   **Ingredient Normalization:** The consistent conversion of ingredients to lowercase (both available and recipe ingredients) correctly handles case-insensitivity, which is crucial for user experience.
*   **Robust CLI Handling:** The `main` function effectively uses `argparse` to process command-line arguments and includes error handling for missing ingredients, exiting gracefully.
*   **Clear Output Formatting:** The console output for recommendations is well-structured and thematic, making it easy for users to understand the results.

🧪 Tests
*   **Comprehensive Unit Tests:** The `test_recommender.py` suite covers a wide range of scenarios, including no matches, partial matches, full matches, case insensitivity, duplicate ingredients, and single-ingredient recipes.
*   **Effective Mocking for CLI:** The use of `@patch` for `argparse.ArgumentParser.parse_args`, `sys.stdout`, and `sys.exit` in `TestMainFunction` is an excellent approach to test the command-line interface without actual side effects, ensuring the `main` function behaves as expected.
*   **Isolated Test Data:** The `setUp` methods in the test classes create specific `SurvivalCookbook` instances with controlled recipe data, preventing test interference and ensuring predictable results.
*   **Actionable Feedback:**
    *   Consider adding a test for edge cases with empty ingredient lists in recipes, although the current implementation implicitly handles this by having non-empty recipe ingredient lists.
    *   While the current `setUp` for `TestSurvivalCookbook` and `TestRecipeRecommendation` is functional, for larger test suites, centralizing the creation of a standard test cookbook state could reduce redundancy and improve clarity.

🔒 Security
*   **Self-Contained Utility:** The utility operates without external network calls or file I/O beyond its own script, significantly limiting potential attack vectors.
*   **No Sensitive Data Handling:** The code does not process or store any sensitive user data, credentials, or secrets.
*   **Standard Library Usage:** Reliance solely on Python's standard library minimizes the risk associated with third-party dependencies.
*   **Actionable Feedback:**
    *   No immediate security vulnerabilities are apparent given the scope and nature of the utility. Continue to ensure that any future enhancements (e.g., loading recipes from external files) implement robust input validation and sanitization to prevent injection or path traversal issues.

🧩 Docs/DX
*   **Clear README:** The `README.md` provides a good overview, features list, and clear instructions for usage and testing, including example output.
*   **Thematic Content:** The "Whimsical Survival Lore" and post-apocalyptic theme are well-integrated and enhance the user experience.
*   **Actionable Feedback:**
    *   **Correct Directory Path in README:** The `How to Use` section's `cd` command specifies an incorrect path. The current directory structure is `utils/nightly-nightly-rad-resistant-recipe/utils/nightly-rad-resistant-recipe-recommender/`. The `cd` command should reflect the full path:
        ```diff
        --- a/utils/nightly-nightly-rad-resistant-recipe/utils/nightly-rad-resistant-recipe-recommender/README.md
        +++ b/utils/nightly-nightly-rad-resistant-recipe/utils/nightly-rad-resistant-recipe-recommender/README.md
        @@ -14,7 +14,7 @@
         1.  **Navigate to the utility directory**:
             ```bash
             -cd utils/nightly-rad-resistant-recipe-recommender
            +cd utils/nightly-nightly-rad-resistant-recipe/utils/nightly-rad-resistant-recipe-recommender
             ```
         ```
    *   **Simplify Directory Structure:** The nested directory `utils/nightly-nightly-rad-resistant-recipe/utils/nightly-rad-resistant-recipe-recommender/` is redundant. Consider flattening it to `utils/nightly-rad-resistant-recipe-recommender/` for better navigability and cleaner paths. This would also simplify the `cd` command in the README.
    *   **Consider External Recipe Data:** While hardcoded recipes are fine for a self-contained utility, for future extensibility or easier content updates, explore options for loading recipes from a simple data file (e.g., JSON, YAML) rather than directly in `_load_default_recipes`. This would separate data from code logic.

🧱 Mocks/Fakes
*   **Appropriate Use in Tests:** Mocks are effectively used in `test_recommender.py` to isolate the `main` function's behavior from actual command-line parsing and standard output, allowing for focused testing of its logic and output.
*   **Clear Rationale:** The comments within the `TestMainFunction` class provide clear rationales for why mocking is being used, which is helpful for understanding the test's intent.
*   **Actionable Feedback:**
    *   The PR body's "Mock Justification: Not applicable; generator did not introduce new mocks" might be slightly misleading. While the *generator* didn't introduce mocks *into the main application code*, the *tests* for the generated code *do* use mocks. Clarify in the PR body that mocks are used extensively and appropriately within the test suite to ensure comprehensive coverage of the CLI interface.

**Comment by polsala** [comment]
## 🤖 Review by GROQ Agent

<think>
Okay, let's start by looking at the PR. The title is "feat(gemini): autogenerated update," and the body mentions an automated proposal from the Gemini generator. The user wants a review in a specific structure with sections like What's solid, Tests, Security, Docs/DX, and Mocks/Fakes. Each section needs actionable feedback with bullet points and code snippets where helpful.

First, the "What's solid" section. The PR has a modular design with classes like Recipe and SurvivalCookbook. The ingredient normalization to lowercase is a good point for case insensitivity. Also, the test coverage seems comprehensive, covering edge cases like no matches, partial matches, and case insensitivity. The README is detailed, which is good for DX.

Next, the Tests section. The test cases cover various scenarios, which is solid. But I notice that the test for the main function uses mocks for argparse and sys.stdout. That's okay, but maybe there's a way to test the actual CLI without mocks, perhaps by using subprocess. Also, the test for no ingredients provided uses a mock to check the error message. That's good, but maybe the error message could be more specific or include examples.

Security-wise, the code doesn't handle any sensitive data, so that's a plus. But since it's a CLI tool, input validation is important. The current code checks for empty ingredients, but what about invalid inputs like numbers or special characters? Also, the code uses argparse, which is safe, but maybe there's a need to sanitize the input ingredients to prevent injection attacks, though that's less likely here.

Docs/DX: The README is good, but the installation instructions are missing. Users might need to know if there are dependencies. Also, the example output in the README shows some recipes, but the actual code in the diff has a truncated recipe (Foraged Berry Mash). That might be an issue if the test cases don't match the actual code. Also, the test cases mention recipes like "Hearty Stew" and "Simple Soup," but those aren't in the provided code snippet. That could be a problem if the tests are relying on recipes that aren't present.

Mocks/Fakes: The tests use mocks for argparse and sys.stdout, which is appropriate. However, the test_main_with_recommendations uses MagicMock for parse_args, which might not capture all edge cases. Maybe using a real argument parsing setup would be better, but mocks are okay here. Also, the test for no ingredients uses a mock to check the exit code, which is correct.

Putting this all together, I need to structure each section with actionable feedback. For example, in Tests, suggest adding more input validation tests. In Docs/DX, mention adding installation instructions. In Security, note the input sanitization. Each point should be a bullet with code examples where applicable.
</think>

```markdown
## ✅ What's solid
- **Modular Design**: The separation into `Recipe`, `SurvivalCookbook`, and `recommend_recipes` promotes testability and clarity.
- **Ingredient Normalization**: Lowercasing ingredients ensures case-insensitive matching (e.g., `Water` → `water`).
- **Test Coverage**: Comprehensive test cases for edge conditions like partial matches, case insensitivity, and empty inputs.
- **README Quality**: Clear usage instructions with example commands and output formatting.

---

## 🧪 Tests
- **Missing Input Validation Tests**:
  - Test for invalid input types (e.g., numbers, special characters in `--ingredients`).
  - Verify handling of malformed CLI arguments (e.g., missing `--ingredients` flag).
- **Test Output Consistency**:
  - The test `test_main_with_recommendations` assumes specific output formatting (e.g., `*Survival Tip:`). If the `Recipe.tip` format changes, tests may break. Consider using regex or dynamic checks instead of hardcoded strings.
  - Example fix for robustness:
    ```python
    # Instead of:
    self.assertIn("*Survival Tip:", output)
    # Use regex to match any tip line:
    self.assertRegex(output, r"\*Survival Tip: .+")
    ```
- **Missing Test for Duplicate Recipes**:
  - Add a test to ensure the same recipe isn't recommended multiple times if ingredients match in different ways.

---

## 🔒 Security
- **Input Sanitization**:
  - The CLI accepts raw user input (`--ingredients`). Consider sanitizing inputs to prevent code injection (e.g., strip unexpected characters).
  - Example fix:
    ```python
    def sanitize_ingredients(ingredients: List[str]) -> List[str]:
        return [re.sub(r'[^a-z0-9 ]', '', ing.lower()) for ing in ingredients]
    ```
- **No Dependency Checks**:
  - The utility uses `argparse` and `unittest` (standard libraries), but if external dependencies are added later, include a `requirements.txt` with pinned versions.

---

## 🧩 Docs/DX
- **Missing Installation Instructions**:
  - The README lacks guidance on installing dependencies (if any). Add a section like:
    ```markdown
    ## 🧰 Installation
    No external dependencies required. Ensure Python 3.8+ is installed.
    ```
- **Recipe Data Inconsistency**:
  - The test cases reference recipes like `Hearty Stew` and `Simple Soup`, but these are not present in the provided `recommender.py` snippet. Verify that all test recipes are defined in `_load_default_recipes()`.
- **Example Output Truncation**:
  - The README's example output includes `Bunker Bean Stew` but the truncated `recommender.py` diff shows a recipe named `Foraged Berry Mash`. Ensure examples match actual code.

---

## 🧱 Mocks/Fakes
- **Over-Reliance on Mocks**:
  - The `test_main_with_recommendations` uses `@patch('argparse.ArgumentParser.parse_args')` to simulate CLI input. Consider testing the actual CLI invocation with `subprocess.run()` for end-to-end validation:
    ```python
    def test_cli_output():
        result = subprocess.run(
            ['python', 'recommender.py', '--ingredients', 'beans water'],
            capture_output=True, text=True
        )
        assert "Bunker Bean Stew" in result.stdout
    ```
- **Mock Exit Code Verification**:
  - The `test_main_no_ingredients_provided` mocks `sys.exit()`. Instead of verifying the exit code via `mock_exit`, explicitly assert the return value of `main()` if refactored to return an exit code.
```
